### ISSUE TITLE
Support exact search in a better way

### ISSUE THREAD

**Issue by benwtrent** [elastic/elasticsearch#97541]
### Description

Currently the only way for users to do a brute-force or exact search with KNN is with a script query. This requires some knowledge of the function names and scoring methodologies. 

We should provide a better interface for exact scan. One idea is to have an `exact: true` field within `kNN`. 

The name of the field is debatable. Or if we even update kNN at all. Maybe a new `kNN` query that allows for both exact and approximate within the query DSL?

**Comment by benwtrent** [comment]
Some edge cases here are:
 - `dense_vector` fields that are not indexed, this means they don't have a similarity_function defined. We need to allow users to specify their expected similarity function AND have a good default defined.
 - Folks that want to use a different similarity function than the one that is indexed in `dense_vector`. Should we allow this? It's technically possible as `exact` will simply be iterating the vectors and calculating similarity.

**Comment by carlosdelest** [comment]
## Task Refinement 
cc @benwtrent and @mayya-sharipova for validation

### Questions

- Is this task to be done before or after [Make knn search as a query](https://github.com/elastic/elasticsearch/issues/97940)? I don't think there's a final conclusion in [this doc](https://docs.google.com/document/d/1YtbOUyIm7CymjpfzXCtiClaEguuJfUvrl7Jpybkbdps/edit#heading=h.qgh2534ub6mk).

### Implementation plan

In case we do this before making knn search a query, these are the changes to the code that I've planned:

#### Lucene

- Modify `AbstractKnnVectorQuery` to add an `exact` attribute to it for doing exact search.
- Modify `AbstractKnnVectorQuery.getLeafResults` to do an exact search in case `exact` attribute is set, similar to the current exact searches that are done due to too many nodes visited or less than k possible matches:

```java
    if (exact) {
      Scorer scorer = filterWeight.scorer(ctx);
      BitSet acceptDocs = createBitSet(scorer.iterator(), liveDocs, maxDoc);

      return exactSearch(ctx, new BitSetIterator(acceptDocs, k));
    }
```

- Update `KnnVectorQuery`, `KnnByteVectorQuery` and `KnnFloatVectorQuery` queries adding the `exact` attribute. I'm thinking on using a Builder or a Parameter Object to avoid duplicating current constructors, thought that would involve changes in the callers as well.
- Add tests for `exact` query parameter and ensuring it performs an exact search. I'm thinking about checking that the proper method (`exactSearch`) is invoked, but happy to hear other opinions on how to test this.

#### Elasticsearch

- Modify `KnnVectorQueryBuilder` to add an `exact` parameter
- Change `KnnVectorQueryBuilder.doToQuery` to pass along the `exact` parameter to `DenseVectorFieldMapper.createKnnQuery`
- Modify `DenseVectorFieldMapper.createKnnQuery` to create the Lucene `KnnByteVectorQuery` or `KnnFloatVectorQuery` with the Elasticsearch query`exact` parameter
- Add YAML tests and `AbstractKnnVectorQueryBuilderTestCase` tests

**Comment by carlosdelest** [comment]
After discussing with @benwtrent , we'll tackle this issue after https://github.com/elastic/elasticsearch/issues/97940

**Comment by carlosdelest** [comment]
This is no longer blocked - @liranabn for prioritisation.

**Comment by benwtrent** [comment]
We need to consider the case when vectors are not in an HNSW graph at all (e.g. "index: false"). We need to allow kNN queries and top level kNN to work there as well IMO. This may require some configuration from the user to indicate the similarity they want to use. Possibly, we just require the similarity to be set in the mapping and if they want to use custom similarity functions, they must switch back to `script`.

I wonder if we should allow `similarity` to be stored in the mapping configuration even when `index: false`.

**Comment by saikatsarkar056** [comment]
From the above discussion, we will take the following steps for the scope of this work.

- Based on this [comment](https://github.com/elastic/elasticsearch/issues/97541#issuecomment-1706175327), create a PR in Lucene.
- Wait for the latest release of lucene and [upgrade of lucene](https://github.com/elastic/elasticsearch/blob/7c6120b50f03088d4fdac0e81994d991c54d20a7/build-tools-internal/version.properties#L2) in elasticsearch.
- Based on this [comment](https://github.com/elastic/elasticsearch/issues/97541#issuecomment-1706175327), create a PR in Elasticsearch.
- Handling the edge cases is out of scope for this issue.

**Comment by saikatsarkar056** [comment]
For the Lucene changes, we need a new public interface that all leaf readers can read. 

Assigning the issue to @benwtrent for the lucene work. Once the lucene work is done, search relevance team can take the elasticsearch work.

**Comment by benwtrent** [comment]
The plan here is to have a new `dense_vector` query. This query will accept some parameters:

 - one to indicate that it allows quantized scores if available
 - another indicating the similarity desired (default to cosine or indexed field)

**Comment by michaeljmarshall** [comment]
@benwtrent - a couple of implementation details I want to call out:
1. The implementation uses `similarity_function` instead of `similarity`.
2. **`bit` + `index: false`** isn't supported. A query that attempts to score these will result in an error directing the user to index the field, since Hamming over packed bits isn't a `VectorSimilarityFunction`.
3. When a query sets `quantized: true` but there are not quantized vectors, we compute the scores using the full precision vectors.

**Comment by benwtrent** [comment]
> bit + index: false isn't supported. A query that attempts to score these will result in an error directing the user to index the field, since Hamming over packed bits isn't a VectorSimilarityFunction.

Part of our goal is to be feature parity with the scoring users can do within our scripting. A key part of that is hamming, which is just squared euclidean (which is what we return already to users, just inverted).

**Comment by michaeljmarshall** [comment]
Thanks for confirming, I'll look into wiring that up too.

**Comment by michaeljmarshall** [comment]
I pushed up a commit to add support for bit vectors. I plan on reviewing the PR in its totality tomorrow, then it should be ready for review.

**Comment by michaeljmarshall** [comment]
Completed via https://github.com/elastic/elasticsearch/pull/147820
